File: python/v1/buyers/user_lists/update_user_lists.py
#!/usr/bin/python
#
# Copyright 2019 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""This example updates a userLists resource with the specified name.

Note that a user list's status can not be adjusted with the update method. See
close_user_lists.py and open_user_lists.py for examples of modifying the
state of existing user lists.
"""


import argparse
import datetime
import os
import pprint
import sys
import uuid

sys.path.insert(0, os.path.abspath('../../..'))

from googleapiclient.errors import HttpError
import util


_USERLISTS_NAME_TEMPLATE = 'buyers/%s/userLists/%s'

DEFAULT_BUYER_RESOURCE_ID = 'ENTER_BUYER_RESOURCE_ID_HERE'
DEFAULT_USERLIST_RESOURCE_ID = 'ENTER_USERLIST_RESOURCE_ID_HERE'


def main(realtimebidding, args):
  account_id = args.account_id
  user_list_id = args.user_list_id

  user_list = {
      'displayName': args.display_name,
      'urlRestriction': {
          'url': args.url,
          'restrictionType': args.restriction_type,
          'startDate': {
              'year': args.start_year,
              'month': args.start_month,
              'day': args.start_day
          },
          'endDate': {
              'year': args.end_year,
              'month': args.end_month,
              'day': args.end_day
          }
      }
  }

  if args.description:
    user_list['description'] = args.description

  print(f'Updating userList "{user_list_id}" for buyer account ID '
        f'"{account_id}":')
  try:
    # Construct and execute the request.
    response = (realtimebidding.buyers().userLists().update(
                name=_USERLISTS_NAME_TEMPLATE % (account_id, user_list_id),
                body=user_list).execute())
  except HttpError as e:
    print(e)
    sys.exit(1)

  pprint.pprint(response)


if __name__ == '__main__':
  try:
    service = util.GetService(version='v1')
  except IOError as ex:
    print(f'Unable to create realtimebidding service - {ex}')
    print('Did you specify the key file in util.py?')
    sys.exit(1)

  def valid_restriction_type(restriction_type):
    _VALID_RESTRICTION_TYPES = [
        'CONTAINS', 'EQUALS', 'STARTS_WITH', 'ENDS_WITH', 'DOES_NOT_EQUAL',
        'DOES_NOT_CONTAIN', 'DOES_NOT_START_WITH', 'DOES_NOT_END_WITH']

    if restriction_type.upper() in _VALID_RESTRICTION_TYPES:
      return restriction_type
    else:
      raise ValueError('Invalid URL restriction type specified. Must be one '
                       f'of: {_VALID_RESTRICTION_TYPES}')

  today = datetime.date.today()

  parser = argparse.ArgumentParser(
      description=('Updates the specified user list.')
  )

  # Required fields.
  parser.add_argument(
      '-a', '--account_id', default=DEFAULT_BUYER_RESOURCE_ID,
      help=('The resource ID of the buyers resource under which the '
            'userlist was created. This will be used to construct the '
            'name used as a path parameter for the userlists.update request.'))
  parser.add_argument(
      '-u', '--user_list_id', default=DEFAULT_USERLIST_RESOURCE_ID,
      help=('The resource ID of the buyers.userlists resource for which the '
            'userlist was created. This will be used to construct the '
            'name used as a path parameter for the userlists.update request.'))
  # Optional fields.
  parser.add_argument(
      '-n', '--display_name', default=f'Test UserList #{uuid.uuid4()}',
      help='The user-specified display name of the user list.')
  parser.add_argument(
      '-d', '--description', default=None,
      help='The user-specified description of the user list.')
  parser.add_argument(
      '--url', default='https://luxurymarscruises.com',
      help='The URL to use for applying the UrlRestriction on the user list.')
  parser.add_argument(
      '-r', '--restriction_type', default='EQUALS',
      type=valid_restriction_type,
      help=('The URL to use for applying the UrlRestriction on the user list. '
            'By default, this will be set to EQUALS. For more details on how '
            'to interpret the different restriction types, see: '
            'https://developers.google.com/authorized-buyers/apis/'
            'realtimebidding/reference/rest/v1/buyers.userLists?hl=en'
            '#UrlRestriction.FIELDS.restriction_type'))
  parser.add_argument(
      '--start_day', default=today.day, type=int,
      help=('Starting day of the month for the URL restriction. If default '
            'values are used, the startDate will be set to today.'))
  parser.add_argument(
      '--start_month', default=today.month, type=int,
      help=('Starting month (numeric) for the URL restriction. If default '
            'values are used, the startDate will be set to today.'))
  parser.add_argument(
      '--start_year', default=today.year, type=int,
      help=('Starting year for the URL restriction. If default values are '
            'used, the startDate will be set to today.'))
  parser.add_argument(
      '--end_day', default=today.day + 1, type=int,
      help=('Ending day of the month for the URL restriction. If default '
            'values are used, the endDate will be set to tomorrow.'))
  parser.add_argument(
      '--end_month', default=today.month, type=int,
      help=('Ending month (numeric) for the URL restriction. If default '
            'values are used, the endDate will be set to tomorrow.'))
  parser.add_argument(
      '--end_year', default=today.year, type=int,
      help=('Ending year for the URL restriction. If default values are used, '
            'the endDate will be set to tomorrow.'))

  args = parser.parse_args()

  main(service, args)

File: java/src/main/java/com/google/api/services/samples/authorizedbuyers/realtimebidding/v1/buyers/userlists/UpdateUserLists.java
/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.api.services.samples.authorizedbuyers.realtimebidding.v1.buyers.userlists;

import com.google.api.services.realtimebidding.v1.RealTimeBidding;
import com.google.api.services.realtimebidding.v1.model.Date;
import com.google.api.services.realtimebidding.v1.model.UrlRestriction;
import com.google.api.services.realtimebidding.v1.model.UserList;
import com.google.api.services.samples.authorizedbuyers.realtimebidding.Utils;
import java.io.IOException;
import java.security.GeneralSecurityException;
import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.ArgumentParserException;
import net.sourceforge.argparse4j.inf.Namespace;
import org.joda.time.format.DateTimeFormatter;

/**
 * This sample illustrates how to update UserLists.
 *
 * <p>Note that a user list's status can not be adjusted with the update method.
 */
public class UpdateUserLists {
  public static void execute(RealTimeBidding client, Namespace parsedArgs) throws IOException {
    Long accountId = parsedArgs.getLong("account_id");
    String userListId = parsedArgs.getString("user_list_id");
    String userListName = String.format("buyers/%s/userLists/%s", accountId, userListId);
    DateTimeFormatter formatter = Utils.getDateTimeFormatterForLocalDate();
    String startDateStr = parsedArgs.getString("start_date");
    String endDateStr = parsedArgs.getString("end_date");
    Date startDate = null;
    Date endDate = null;

    if (startDateStr != null) {
      startDate = Utils.convertJodaLocalDateToRTBDate(formatter.parseLocalDate(startDateStr));
    }

    if (endDateStr != null) {
      endDate = Utils.convertJodaLocalDateToRTBDate(formatter.parseLocalDate(endDateStr));
    }

    UserList origUserList = client.buyers().userLists().get(userListName).execute();

    UrlRestriction origUrlRestriction = origUserList.getUrlRestriction();

    String displayName = parsedArgs.getString("display_name");
    if (displayName != null) {
      origUserList.setDisplayName(displayName);
    }

    String description = parsedArgs.getString("description");
    if (description != null) {
      origUserList.setDescription(description);
    }

    String url = parsedArgs.getString("url");
    if (url != null) {
      origUrlRestriction.setUrl(url);
    }

    String restrictionType = parsedArgs.getString("restriction_type");
    if (restrictionType != null) {
      origUrlRestriction.setRestrictionType(restrictionType);
    }

    if (startDate != null) {
      origUrlRestriction.setStartDate(startDate);
    }

    if (endDate != null) {
      origUrlRestriction.setEndDate(endDate);
    }

    Long membershipDurationDays = parsedArgs.getLong("membershipDurationDays");
    if (membershipDurationDays != null) {
      origUserList.setMembershipDurationDays(membershipDurationDays);
    }

    origUserList.setUrlRestriction(origUrlRestriction);

    UserList updatedUserList =
        client.buyers().userLists().update(userListName, origUserList).execute();

    System.out.printf("Updated UserList for buyer Account ID '%s':\n", accountId);
    Utils.printUserList(updatedUserList);
  }

  public static void main(String[] args) {
    ArgumentParser parser =
        ArgumentParsers.newFor("CreateUserLists")
            .build()
            .defaultHelp(true)
            .description(("Updates the specified user list."));
    parser
        .addArgument("-a", "--account_id")
        .help(
            "The resource ID of the buyers resource under which the user list was created. "
                + "This will be used to construct the name used as a path parameter for the "
                + "userlists.update request.")
        .required(true)
        .type(Long.class);
    parser
        .addArgument("-u", "--user_list_id")
        .help(
            "The resource ID of the buyers.userlists resource for which the user list was created."
                + " This will be used to construct the name used as a path parameter for the"
                + " userlists.update request.")
        .required(true)
        .type(Integer.class);
    parser
        .addArgument("-n", "--display_name")
        .help("The user-specified display name of the user list.");
    parser
        .addArgument("-d", "--description")
        .help("The user-specified description of the user list.");
    parser
        .addArgument("--url")
        .help("The URL to use for applying the UrlRestriction on the user list.");
    parser
        .addArgument("-r", "--restriction_type")
        .help(
            "The restriction type for the specified URL. For more details on how to interpret the"
                + " different restriction types, see the reference documentation: "
                + "https://developers.google.com/authorized-buyers/apis/realtimebidding/reference/rest/"
                + "v1/buyers.userLists#UrlRestriction.FIELDS.restriction_type")
        .choices(
            "CONTAINS",
            "EQUALS",
            "STARTS_WITH",
            "ENDS_WITH",
            "DOES_NOT_EQUAL",
            "DOES_NOT_CONTAIN",
            "DOES_NOT_START_WITH",
            "DOES_NOT_END_WITH");
    parser
        .addArgument("--start_date")
        .help(
            "The start date for the URL restriction, specified as an ISO 8601 date "
                + "(yyyy/mm/dd. By default, this will be set to today.");
    parser
        .addArgument("--end_date")
        .help(
            "The end date for the URL restriction, specified as an ISO 8601 date "
                + "(yyyy/mm/dd. By default, this will be set to tomorrow.");
    parser
        .addArgument("-m", "--membership_duration_days")
        .help(
            "The number of days a user's cookie stays on the user list. The value must be "
                + "between 0 and 540 inclusive. This will be set to 30 by default.")
        .type(Long.class);

    Namespace parsedArgs = null;
    try {
      parsedArgs = parser.parseArgs(args);
    } catch (ArgumentParserException ex) {
      parser.handleError(ex);
      System.exit(1);
    }

    RealTimeBidding client = null;
    try {
      client = Utils.getRealTimeBiddingClient();
    } catch (IOException ex) {
      System.out.printf("Unable to create RealTimeBidding API service:\n%s", ex);
      System.out.println("Did you specify a valid path to a service account key file?");
      System.exit(1);
    } catch (GeneralSecurityException ex) {
      System.out.printf("Unable to establish secure HttpTransport:\n%s", ex);
      System.exit(1);
    }

    try {
      execute(client, parsedArgs);
    } catch (IOException ex) {
      System.out.printf("RealTimeBidding API returned error response:\n%s", ex);
      System.exit(1);
    } catch (IllegalArgumentException ex) {
      System.out.printf(
          "The specified start_date (%s) or end_date (%s) arguments could not " + "be parsed:\n%s",
          parsedArgs.getString("start_date"), parsedArgs.getString("end_date"), ex);
      System.exit(1);
    }
  }
}
